Added DateRangePicker

tungi52 opened the pull request:
According to #309 I started to implement a DateRangePicker. I don't want to break the current design, so I tried to enhance the actual implementation and keep common features together.
What is changed:
- new RangeInput component, which can accepts two values - this can be used in the future for other range inputs (numeric or time)
- new DateRangePicker component
- a few objects, necessary to achieve the required behaviour  

Please check it, I appreciate your opinions.

tungi52 committed:
added date-range picker

css updates

mikes-gh commented:
@tungi52 Nice work. 😄  I ran your branch and liked it.  The code looks of good quality too.
A couple of things that might improve it would be some css that makes the range look more continous like https://flatpickr.js.org/examples/#range-calendar
Some tests if accepted.

tungi52 commented:
@mikes-gh Thank you very much! Yes I agree, the css is not nice at the moment, currently I wanted to share the code with you, guys. I will check the link and work on it. I will add some tests too before mark as "ready for review".

mikes-gh commented:
@tungi52 Please note I am not a maintainer so its just my opinion.

tungi52 commented:
@mikes-gh I appreciate your opinion and I take your advices.

Garderoben commented:
I like it! the textfields needs some work but i'm not sure how much work that would be do fix? (im redoing the whole textfield structure atm)

I would like to see it in the same component and have the toolbar in there as well, kinda like Vuetify has it, but instead of "2 selected" maybe do the dates? 01-27 or something?

![image](https://user-images.githubusercontent.com/10367109/103848111-ee208f00-50a1-11eb-861a-8f9122b427fd.png)

tungi52 commented:
@Garderoben @mikes-gh Thanks for your feedback, I'm going to improve the component based on your suggestions.

mikes-gh commented:
css look much better 😄 
- When I select the first date I feel trapped because if I selected say the 6th I cant reslect 5th without selecting an end date first.
- Also the root of your branch is getting old.  I know you said git wasnt your thing but you need to rebase your branch.  You are going to have some conflicts to resolve as I have done a few fixes to the date picker (sorry they couldnt wait).

mikes-gh commented:
I can guide you on the rebase.  Its a vital skill to learn for contributing.

tungi52 commented:
@mikes-gh Yes, I noticed your changes in the datepicker yesterday, so I tried to rebase my branch. I thought I succeeded :( There was only a few changes in your PR, so I tracked them manually, and my branch seemed ok for me. Or did I miss something?

mikes-gh commented:
You merged develop into your feature branch which is not a rebase.  While it does the job - for branches that are yours it is better to do a rebase as it keeps the commit history clean.  see https://www.atlassian.com/git/tutorials/merging-vs-rebasing

tungi52 committed:
minor fixes+tests

tungi52 commented:
@mikes-gh I tried to rebase my branch, can you check it? 
Also I made some fixes and add tests based on the DatePicker test.

mikes-gh commented:
@tungi52 looks like you got the hang of it 👍 .  Lets review after your position fix as you will need to rebase again.  Then things will be clearer and we can test at the HEAD of develop.

mikes-gh commented:
@tungi52 I am really liking this. 👍   If I add ideas its just because I like what you're doing.  Is it possible to have `n months` open to make selecting across months easier.  Could be for the future but good to think about it early in the design phase.  I have always liked [this by Dan Grossman](http://www.daterangepicker.com/).

mikes-gh commented:
Another thing on selecting the second date before the first reset the selection would seem the most natural IMO.  Covers the situation where mis click/tap the first date

mikes-gh commented:
More ideas in reporting your report might need a range of one day.  In which case a second click on the start date should make the date round css and return for example  31/12/21 -> 31/12/21

tungi52 commented:
@mikes-gh no problem, I always ready for new ideas and I like different perspectives. I'm going to implement what you suggested.
About the displaying more than one month: I don't think it is impossible, the current MudBaseDatePicker should be improved. But I think we should aware the screen size (on small screens the months should be aligned vertically). Also, the alignment could be a parameter... I think I'll give it a try.

mikes-gh commented:
> But I think we should aware the screen size

Yes I think there could do with a compact /condensed version later on.  Also on iPad etc a swipe for month change would be good.  Ive got loads of ideas because the business reporting apps  I usually write are date range orientated.  But I think you are close to a v1 of this.  Better to have a v1 than none at al IMO.

tungi52 committed:
improvement+n-month-display+weeknumber

tungi52 committed:
styling, examples and test

tungi52 commented:
@mikes-gh I changed the range picker behaviour as you suggested: 1 day range, and reset selection by clicking a date smaller than the previous click.
Also I tried to implement an n month view, if you have time, can you take a look at it? I am interested in your opinion.
Additionally I add a new feature: datepicker can display the weeknumber now. As I experienced, it is a very common request.
I updated to docs and test according to the changes.

mikes-gh commented:
@tungi52 I like it again 😄  I think the number of calendars should be configureable but apart from that I think its a good v1.

tungi52 commented:
@mikes-gh Thank you! It is configurable already, I added a `DisplayMonth` parameter.
@Garderoben @mikes-gh If you agree, I think this is ready for review.

mikes-gh commented:
Sure but DisplayMonth sounds like it should be set to a single month?  I think you need NoOfMonthsToDisplay ? Or somethiing like that.

mikes-gh commented:
and an offset for monthtodsplay so you could for example open a month before and a month after current range of current date

mikes-gh commented:
But I feel like Im getting picky now.  The property name def neeeds changing though.  Great job 👍

tungi52 commented:
@mikes-gh No, no, it's ok. Yes plural, my mistake :) Should be DisplayMonth**s**, I missed that.
I'm not sure I understand MonthsToDisplayOffset. Should it be valid if I want to show 2 months, but I set -3, for example? Or it must be `0 ... (MonthsToDisplay-1)` ? And what about positive values?

mikes-gh commented:
@tungi52 The offset is about what month to start on.  For example you might want to display the previous month as the starting month.

mikes-gh commented:
You could use StartMonth as a DateTime and ignore the days.  Then if you want display a previous month as a start month you can use `DateTime.Now.AddMonths(-1);`

tungi52 committed:
Add start month

tungi52 commented:
@mikes-gh I updated my code. Is what you think about?

mikes-gh commented:
Hi @tungi52 I just ran it and I noticed the all the static calenders on the examples page are blank.

tungi52 committed:
fix static picker

tungi52 commented:
@mikes-gh I fixed it.

mikes-gh commented:
Great.  Looking good.   Sorry noticed one other thing -  week nos css need some work.  Im no designer But it doesn't look right and  probably needs a column title where days names are `WkNo` or something.  Maybe centre justify and give them a bit of margin.

<img width="133" alt="Screenshot 2021-01-10 at 12 37 32" src="https://user-images.githubusercontent.com/16208742/104123073-d450a780-5340-11eb-9c7f-43b582f32188.png">

mikes-gh commented on src/MudBlazor/Components/DatePicker/MudBaseDatePicker.razor.cs:
```diff
@@ -104,7 +104,7 @@ public abstract partial class MudBaseDatePicker : MudBasePicker
         protected virtual bool IsRange { get; } = false;
 
-        private OpenTo _currentView;
+        private OpenTo _currentView = OpenTo.Date;
```
I doubt this is the right fix.  Please check why OpenTo is not being copied in this case

mikes-gh replied:
What if we want a year static or a month static?

tungi52 replied:
Yes. The problem was earlier, that OpenTo enum got a None value, and this is the default. As you said, OnPickerOpened doesn't called when picker is static, so year and month start option I think currently not working at all. Copying `_currentView = OpenTo;` firstrender is not a solution too, because in dialog mode after closing the picker, it will not return to year/month view on the second open. Maybe we should handle static mode separately in firstrender,  like:
```

if (PickerVariant==PickerVariant.Static)
  _currentView = OpenTo;
```

mikes-gh replied:
In the TimePicker I do this in OnInitialised.  Not sure if it makes any difference.

tungi52 replied:
I think there is a slightly small difference, but to be consistent, I changed my code.

mikes-gh replied:
Loading the default in OnItalised seems the best .  You dont have to worry about the inline version as Open event will reset as needed.  It also menas you dont have to cal StateHasChanged

```
        protected override void OnInitialized()
        {
            _currentView = OpenTo;
        }
```
You could probably move the month defaults there too

tungi52 replied:
You mean this line: `_picker_month = StartMonth?.StartOfMonth() ?? GetMonthStart(0);` ?

mikes-gh reviewed (commented):
please could you check _currentView should be copying from OpenTo on open.  Manipulating the field directly I dont think is right.  Just realised OnOpen doesnt happen for static.  Maybe first Render?

tungi52 commented:
@mikes-gh Yes, I see that one pixel difference now, I haven't centered the weeknumber entirely. I haven't gave a title for it on purpose, because I think it could be problematic to handle different languages, so the only thing could be universal enough (and not long) is '#'.  Anyway, I used [ng-bootstrap](https://ng-bootstrap.github.io/#/components/datepicker/overview) as reference, and there they also not gave any title to weeknumber.

tungi52 commented:
@mikes-gh Oh, I only see the bug on your picture now. IT must be something with the locale, because it's perfect for me. I need to test it with different cultures.

tungi52 committed:
fixes and styling

mikes-gh commented:
I think this is a great addition thank you for listening to all the feedback. @henon @Garderoben

tungi52 committed:
Update MudBaseDatePicker.razor.cs

mikes-gh commented:
JUst saw you read my mind ;smile:  On reflection you probabaly cant move the month stuff there if it can be set dynamically.

mikes-gh commented:
@tungi52 I would make this as ready for review 👍

tungi52 commented:
> @tungi52 I would mark this as ready for review 👍

@mikes-gh Ok, Thank you very much for your help!
